### PR TITLE
Feature/points count

### PR DESCRIPTION
**Pull request by Mapet13**

### CODE REVIEW

File: src/simulation/topology.py [hunk 1]
```diff
@@ -65,19 +73,46 @@ def is_land(binary_map: BinaryMap, top_left_offset: CoordinatesBase[int], x: int
...
+    logger.debug("STATED: Mapping binary lands")
+    top_left_binary_map_offset = get_top_left_offset()
+    
+    def is_land(coord: Coord_t) -> bool:
+        center = Coordinates(
+            latitude=const.point_lat_centers[coord[1]],
+            longitude=const.point_lon_centers[coord[0]]
+        )
+        binary_map_coordinates = project_binary_map_coordinates(center)
+        lat = binary_map_coordinates.longitude - top_left_binary_map_offset.longitude
```
Comment (Mapet13) [inline src/simulation/topology.py]: error in names

File: src/constatnts.py [hunk 2]
```diff
@@ -44,32 +50,45 @@ def set_simulation_coordinates_parameters(top_coord: float,
...
+    width = coordinates_distance(middle_coord_lon(left_coord), middle_coord_lon(right_coord))
+    
+    get_points_count = lambda size: int(ceil(size / Constants.point_side_size))
+    
+    Constants.point_side_lat_count = get_points_count(height)
+    Constants.point_side_lon_count = get_points_count(width)
+    
+    logger.debug(f"Points count: {Constants.point_side_lat_count} x {Constants.point_side_lon_count}")
+
+    def get_center(first, direction, index) -> float:
```
Comment (Herdran) [inline src/constatnts.py]: Type hints for parameters

**Review by Herdran — APPROVED** [review]

File: src/checkpoints.py [hunk 3]
```diff
@@ -25,17 +31,24 @@ def _point_to_dict(point: Point) -> Dict[str, any]:
 
 
 def _oil_source_to_dict(source: Tuple[Coord_t, int, pd.Timestamp, pd.Timestamp]) -> Dict[str, any]:
+    x, y = source[0]
+    lon, lat = get_coordinate_from_xy(x, y).as_tuple()
```
Comment (bblachut) [inline src/checkpoints.py]: Why is it done different way than in function for point above?

File: src/checkpoints.py [hunk 4]
```diff
@@ -25,17 +31,24 @@ def _point_to_dict(point: Point) -> Dict[str, any]:
...
         "spill_end": str(source[3])
     }
 
 
-def save_to_json(world: Dict[Coord_t, Point], total_time: int, curr_iter: int, constant_sources: List[Tuple[Coord_t, int, pd.Timestamp, pd.Timestamp]]) -> None:
+def _get_path_to_save(curr_iter: int) -> PathLike:
     timestamp = time.strftime("%Y_%m_%d-%H_%M_%S")
-    path = get_main_path().joinpath(f"checkpoints/checkpoint_{timestamp}_iteration{curr_iter}.json")
+    checkpoint_dir_path = get_checkpoint_dir_path()
+    checkpoint_dir_path.mkdir(parents=True, exist_ok=True)
```
Comment (bblachut) [inline src/checkpoints.py:45]: Why "parents=True"? If parents don't exists something is definitely not right

File: src/gui/main_screen.py [hunk 5]
```diff
@@ -386,12 +390,13 @@ def update_image_array(self):
                 points_removed = engine.update(self.iter_as_sec)
 
                 for coords in engine.world:
-                    point = engine.world[coords]
-                    if point.topography == simulation.TopographyState.LAND:
-                        var = blend_color(InitialValues.LAND_WITH_OIL_COLOR, InitialValues.LAND_COLOR, point.oil_mass / self.minimal_oil_to_show)
-                    else:
-                        var = blend_color(InitialValues.OIL_COLOR, InitialValues.SEA_COLOR, point.oil_mass / self.minimal_oil_to_show)
-                    self.full_img.putpixel((coords[0], coords[1]), var)
+                    if self.viewer.top <= coords[1] <= self.viewer.bottom and self.viewer.left <= coords[0] <= self.viewer.right:
```
Comment (bblachut) [inline src/gui/main_screen.py]: Shouldn't the second comparison be < instead of  <= for both x and y?

File: src/initial_values_loader.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+from data.generic import Range
+from data.measurement_data import Coordinates
+from data.simulation_run_parameters import Interpolation_grid_size, SimulationRunParameters
+from topology.binary_map_math import project_binary_map_coordinates
+from topology.math import get_xy_dist_from_coord
+
+import pandas as pd
+
+
+logger = getLogger("constants")
```
Comment (bblachut) [inline src/initial_values_loader.py]: initial_values not constants

File: src/log_config.conf [hunk 7]
```diff
@@ -9,20 +9,32 @@ keys=short, all
...
-level = DEBUG
+level = INFO
 
 [logger_data]
 handlers = console, file
 level = DEBUG
 qualname = data
 propagate = 0
 
+[logger_constants]
```
Comment (bblachut) [inline src/log_config.conf]: initial_values not constants

File: src/topology/math.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    full_distance = sqrt((x_distance ** 2) + (y_distance ** 2))
+    return move_coordinate_bearing(InitialValues.top_left_coord, full_distance, degree_bearing)
+
+BEARING_OFFSET = 90.0
+def get_xy_dist_from_coord(first: Coordinates, second: Coordinates) -> (int, int):
+    bearing, distance = calculate_compass_bearing_and_dist(first, second)
+    rad = radians(bearing - BEARING_OFFSET)
+    x = int(distance * cos(rad))
+    y = int(distance * sin(rad))
+    return (x, y)
```
Comment (bblachut) [inline src/topology/math.py]: Remove parenthesis

File: src/topology/math.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    x, y = get_xy_dist_from_coord(InitialValues.top_left_coord, Coordinates(latitude=lat, longitude=lon))
+    x //= InitialValues.point_side_size
+    y //= InitialValues.point_side_size
+    return min(x, InitialValues.point_side_lon_count), min(y, InitialValues.point_side_lat_count)
+
+_geodesic = proj.Geod(ellps='WGS84')
+def calculate_compass_bearing_and_dist(start: Coordinates, end: Coordinates) -> (float, float):
+    forward_azimuth, _, dist = _geodesic.inv(start.longitude, start.latitude, end.longitude, end.latitude)
+    bearing = forward_azimuth + 360 if forward_azimuth < 0 else forward_azimuth
+    return (bearing, dist)
```
Comment (bblachut) [inline src/topology/math.py]: Remove parenthesis

File: src/topology/math.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    point = geo.distance(meters=distance).destination(source.as_tuple(), bearing)
+    return Coordinates(latitude=point.latitude, longitude=point.longitude)
+
+def move_coordinate(source: Coordinates, distance: float, direction: Move_direction) -> Coordinates:
+    return move_coordinate_bearing(source, distance, direction.value)
+
+def move_coordinate_raw(lat: float, lon: float, distance: float, direction: Move_direction) -> (float, float):
+    bearing = direction.value
+    point = geo.distance(meters=distance).destination((lat, lon), bearing)
+    return (point.latitude, point.longitude)
```
Comment (bblachut) [inline src/topology/math.py]: Remove parenthesis

File: src/topology/binary_map_math.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+def _project_coordinates(lat: float, lon: float, width: int, height: int) -> (int, int):
+    x = (lon + LONGITUDE_OFFSET) * (width / LONGITUDE_RANGE)
+    y = (-lat + LATITUDE_OFFSET) * (height / LATITUDE_RANGE)
+    return (int(x), int(y))
+
+def _project_to_coordinates_raw(x: int, y: int, width: int, height: int) -> (float, float):
+    lon = x / (width / LONGITUDE_RANGE) - LONGITUDE_OFFSET
+    lat = -(y / (height / LATITUDE_RANGE) - LATITUDE_OFFSET)
+    return (lon, lat)
```
Comment (bblachut) [inline src/topology/binary_map_math.py]: Remove parenthesis

File: src/topology/binary_map_math.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+LONGITUDE_OFFSET = 180.0
+LATITUDE_OFFSET = 90.0
+LONGITUDE_RANGE = 360.0
+LATITUDE_RANGE = 180.0
+
+def _project_coordinates(lat: float, lon: float, width: int, height: int) -> (int, int):
+    x = (lon + LONGITUDE_OFFSET) * (width / LONGITUDE_RANGE)
+    y = (-lat + LATITUDE_OFFSET) * (height / LATITUDE_RANGE)
+    return (int(x), int(y))
```
Comment (bblachut) [inline src/topology/binary_map_math.py]: Remove parenthesis

File: src/data/measurement_data.py [hunk 13]
```diff
@@ -1,14 +1,19 @@
...
 import numpy as np
 
+
 @dataclass
 class CoordinatesBase(Generic[GenericT]):
     latitude: GenericT
     longitude: GenericT
+    
+    def as_tuple(self) -> tuple[GenericT, GenericT]:
+        return (self.latitude, self.longitude)
```
Comment (bblachut) [inline src/data/measurement_data.py]: Remove parenthesis

File: src/topology/math.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+from data.measurement_data import Coordinates
+
+import geopy.distance as geo
+import pyproj as proj
+
+from initial_values import InitialValues
+
+
+class Move_direction(Enum):
```
Comment (bblachut) [inline src/topology/math.py]: CamelCase

File: src/topology/math.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    return geo.distance(first.as_tuple(), second.as_tuple()).meters
+
+def move_coordinate_bearing(source: Coordinates, distance: float, bearing: float) -> Coordinates:
+    point = geo.distance(meters=distance).destination(source.as_tuple(), bearing)
+    return Coordinates(latitude=point.latitude, longitude=point.longitude)
+
+def move_coordinate(source: Coordinates, distance: float, direction: Move_direction) -> Coordinates:
+    return move_coordinate_bearing(source, distance, direction.value)
+
+def move_coordinate_raw(lat: float, lon: float, distance: float, direction: Move_direction) -> (float, float):
```
Comment (bblachut) [inline src/topology/math.py]: In return type hints use Tuple[], applicable everywhere I posted comment about redundant parenthesis

**Review by bblachut — CHANGES_REQUESTED** [review]

File: src/checkpoints.py (hunk 3 above)
Comment (Mapet13) [reply src/checkpoints.py]: get_coordinate was intended to be point's internal method (and yes, I forgot to use _ sign) that caches results. 
But actually that could be also used here, I will change my initial assumption.

File: src/checkpoints.py (hunk 4 above)
Comment (Mapet13) [reply src/checkpoints.py:45]: > If parents is true, any missing parents of this path are created as needed; they are created with the default permissions without taking mode into account (mimicking the POSIX mkdir -p command).

^ from https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir

File: src/gui/main_screen.py (hunk 5 above)
Comment (Mapet13) [reply src/gui/main_screen.py]: @Herdran

**Review by bblachut — APPROVED** [review]

**Pull request merged**